pandas/core/strings.py

```diff
@@ -2280,6 +2280,23 @@ def cat(self, others=None, sep=None, na_rep=None, join=None):
                                  'must all be of the same length as the '
                                  'calling Series/Index.')
 
+        # data has already been checked by _validate to be of correct dtype,
+        # but others could still have Series of dtypes (e.g. integers) which
+        # will necessarily fail in concatenation. To avoid deep and confusing
+        # traces, we raise here for anything that's not object or all-NA float.
+        def _legal_dtype(series):
```

Comment:
this is way complicated, what exactly are you trying to do here

Reply:
I originally had an inline condition within `any`, but @simonjayhawkins found this too complex, so I broke out that condition into a function.

Basically, I want to fail early for any Series that will necessarily fail concatenation (based on dtype). Object must obviously be allowed, but also all-NA float (which can happen if two Series completely misalign), plus needs handling of categorical.

Reply:
do you not already have the inferred types at this point? and if you don't, why not just infer them, then this condition becomes easier

Reply:
> I originally had an inline condition within `any`, but @simonjayhawkins found this too complex, so I broke out that condition into a function.

the any is applied to a generator expesssion with a for loop and then raising. so i'm not sure the use of any here is a benefit. could you not just use a for loop and do away with the separate function?